Add localisation support

ph1p opened the pull request:
Hey folks,

I've created one type of language support and I've seen the **i18next** branch to late. 
This is why I used another library.

I used react-intl and wrote small spartanic helper functions for main process files.
All languages are located in /locales/*.json.
There is no specific reason why I've used react-intl instead of any other i18n lib.

My main goal was to change as little code as possible and to add new languages easily.

**Menu item**
<img width="364" alt="screen shot 2017-11-12 at 22 13 49" src="https://user-images.githubusercontent.com/15351728/32703475-cbfca9ae-c7f6-11e7-9ae2-b9c125ab3012.png">

**Some links**
- [Post from the creator of i18next](https://github.com/i18next/react-i18next/issues/303)
- [react-intl](https://github.com/yahoo/react-intl)

**Issues**
#285, #292, #284

Closes #285

sallar commented:
Thanks @ph1p for the great work. I quite like this. It's ok to use react-intl as far as I can see, since obviously its easy to use the translations in non-react files too.
Before I review the code, could you check why are the tests failing?

ph1p commented:
@sallar Thank you! I have a problem with the current version of buttercup. I can't run `npm run build:renderer`. This is why I updated all outdated packages (That did not help). 
Webpack freezes at `91% additional asset processing` and node throws **FATAL ERROR: CALL_AND_RETRY_LAST**. 
Your last working commit was `8732150474e9bf6be7cb78faf9bff7f17d9049e2`. 
And the problem is babili (babel-minify). I replaced it with uglify and commented out babel-minify. Now it works, hopefully :)

sallar commented on config/webpack.config.production.js:
```diff
@@ -47,7 +70,55 @@ module.exports = merge(baseConfig, {
       'process.env.NODE_ENV': JSON.stringify('production')
     }),
-    new BabiliPlugin(),
+    new UglifyJSPlugin({
```
Does uglifyJS support ES6 which we are using? I think the reason I used Babili was that it was the only one that supported ES6 out of the box.

ph1p replied:
Jep. Documentation says, that i can use the `ecma` option (https://github.com/webpack-contrib/uglifyjs-webpack-plugin#uglifyoptions)

sallar commented on config/webpack.config.production.js:
```diff
@@ -47,7 +70,55 @@ module.exports = merge(baseConfig, {
+        }
+      }
+    }),
+    // new MinifyPlugin({
```
Can you remove this unused comments?

ph1p replied:
Sure

sallar commented on package.json:
```diff
@@ -108,36 +108,37 @@
   "dependencies": {
-    "opencollective": "^1.0.3"
+    "opencollective": "^1.0.3",
+    "react-intl": "^2.4.0"
```
Can we move this to devDependencies? Since it will be compiled to the final package

ph1p replied:
Jep

sallar commented on src/main/menu.js:
```diff
@@ -11,145 +11,148 @@ import { openFile, openFileForImporting, newFile } from './lib/files';
+import i18n from '../shared/i18n';
 import pkg from '../../package.json';
 
-const defaultTemplate = [
```
Is this diff block because you moved this constant to the setup function?

ph1p replied:
Yes, because I need the `store` variable for all menu points.

sallar commented on src/renderer/index.js:
```diff
@@ -61,10 +63,17 @@ ipc.on('will-quit', () => {
 
+// show message, when locale changed
+ipc.on('locale-changed', (e, payload) => {
+  alert(payload);
```
Can you use the `showDialog` function in `renderer/system/dialog.js`?

sallar replied:
Also, why does the user need to relaunch the program? wont react-intl change the texts in place?

ph1p replied:
Jep

ph1p replied:
Because I can't find a solution to update the whole electron menu.

sallar replied:
@ph1p We do that when a new archive is added. For example I call it everytime the store gets updated: `store.subscribe(debounce(() => setupMenu(store), 500));`. So this should also update the language already if everything's setup correctly

ph1p replied:
Ok cool. I'll try it :)

ph1p replied:
Yeah, It works!

sallar commented on src/shared/i18n/index.js:
```diff
@@ -0,0 +1,104 @@
+  const translations = [];
+
+  config.availableLanguages.forEach(lang => {
+    translations[lang.code] = require(`../../../locales/${lang.code}`);
```
I think this wont work when compiled. Since the paths will change. in `menu.js` I do this for resolving: ```path.resolve(__dirname, `../resources/icons/${item.icon}.png`)``` and I think this should be similar.

sallar replied:
Also can we use `.map` here? 

``` js
const translations = config.availableLanguages.map(lang => require(...));
```

ph1p replied:
It will compile. I've made the forEach, because if you dynamically require content, webpack typically ignores it. With this little trick I require all translations and load them into the object. So when I add a language to the config object, the translations are automatically added.
Anyways, I have updated my code and decided to import the languages by hand.

perry-mitchell commented on locales/de.json:
```diff
@@ -0,0 +1,21 @@
+  "system.dialog.confirm": "Bestätigen",
+  "system.dialog.nevermind": "Abbrechen",
+  "system.dialog.password": "Passwort",
+  "main.menu.language": "Sprache",
```
I really like the nested style, though I think some components should be named differently. For instance, `main.menu` alludes to there being a `main` category with a `menu` subcategory, which I don't think there is. Should this be `main-menu` perhaps? Or even `menu.main`?

perry-mitchell commented on locales/en.json:
```diff
@@ -0,0 +1,21 @@
+  "system.dialog.confirm": "Confirm",
+  "system.dialog.nevermind": "Nevermind",
+  "system.dialog.password": "Password",
+  "main.menu.language": "Language",
```
And here as well..

perry-mitchell commented on locales/es.json:
```diff
@@ -0,0 +1,21 @@
+  "system.dialog.confirm": "Confirmar",
+  "system.dialog.nevermind": "No importa",
+  "system.dialog.password": "Contraseña",
+  "main.menu.language": "Idioma",
```
And here..

perry-mitchell commented:
Also, this shouldn't be merged until we've had a review of the Spanish and German content 😊

ph1p commented:
I'm german, so I reviewed the german part by myself ^^.  
I only added spanish language, because of ticket #284. 
A girlfriend of mine speaks spanish, she will translate it.
These language files are only examples. I'll write a little explanation about the naming and update all of them.
---
Thank you for reviewing my code. I give my best to solve all problems.

ph1p commented:
I've updated my code and have decided to hold the naming of language ids easier. 
The id describes or is itself the translated word written in lower kebab-case.
I think it's better, because it's shorter and easier to use with duplicate words.

perry-mitchell commented:
@ph1p That's awesome, and we appreciate it. As a matter of habit I would personally like all translations (first time) verified by at least 1 other person - it's not a matter of trust but more of habit. I can probably have that done today anyhow 😊

perry-mitchell commented:
Also @sallar - Seeing as this PR adds language **detection**, would it be best to pre-release this to someone running under German/Spanish locale by default?

perry-mitchell commented on src/main/menu.js:
```diff
@@ -204,6 +215,28 @@ export function setupMenu(store) {
+                id: 'language'
+              }),
+              submenu: i18n.getConfig('availableLanguages').map(lang => ({
+                label: lang.name,
```
Is this showing lowercase languages in the menu? Perhaps we could install `ucfirst` to transform the first char to uppercase? What do you think @sallar?

perry-mitchell replied:
Nevermind, I saw that they're capitalised in a later file.

perry-mitchell commented:
I've had the German verified. The Spanish may get a PR later today, but I'm happy with it being merged. Thanks @ph1p !

ph1p commented:
> @ph1p That's awesome, and we appreciate it. As a matter of habit I would personally like all translations (first time) verified by at least 1 other person - it's not a matter of trust but more of habit. I can probably have that done today anyhow 😊

Oh and thanks again for the hard work! This was a massive task, and we're just being extra careful with it. Because of this work, other translations will be much easier 🎉

Thanks again (: I totally understand that.

> Also, @sallar + @ph1p - Are we able to verify what happens if a property in one language is missing? I was hoping that it might default to English, but I can't see yet from the code what the behaviour might be if a translation is not present in one locale..

I'll implement a fallback solution.

perry-mitchell commented:
> I'll implement a fallback solution.

Is that needed? I was just wondering what the behaviour of react-intl was when say a translation exists in English but not in Spanish - Does it fall back automatically?

ph1p commented:
> Does it fall back automatically?

Unfortunately not :/ There is an extra prop for this called `defaultMessage`. I think I'm ready to push it in about an hour.

perry-mitchell commented:
@sallar @ph1p Seeing as this is already functional, do you think we could do without the fallback? Perhaps @ph1p could add it in a separate pull request?

I'd love to see this get merged.. I have a feeling that this will open the flood gates on translations 😍

ph1p commented:
I've added a full translation @perry-mitchell + some bug fixes.

perry-mitchell commented:
> I can't find the word Trash

Ah.. well this is more complex. `General` and `Trash` are groups - content within the Archive data. These are created as boilerplate items in `buttercup-core`. I agree that these should be translated - I can add an option to the core for this, but it should be done later imo (not part of this PR).

ph1p commented:
@perry-mitchell Ah ok and take look at password generator, this must also be dynamic.

perry-mitchell commented:
> password generator, this must also be dynamic

Agreed.. I think we need to figure out a nice way to pass translations around. I guess the generator, being a separate component, should have it's own internationalisation, but should take the current language from the host application (ie. the desktop application in this case).

perry-mitchell commented:
Hmm.. appveyor is still dying. Issue seems to be with `npm`. @ph1p would you mind trying to add `npm cache verify` to the appveyor build script? Before `npm install`?

ph1p commented:
Ok guys ...  I've tried several things, but I don't know why travis and veyor failing. There're stopping at `ava && npm run lint`. I've tested it on my own and can't reproduce it.

sallar commented:
@perry-mitchell @ph1p I'll have a crack at the app veyor thing to see whats dying.

sallar commented:
@ph1p Im going to do some commits to your repo if you dont mind. including reverting some of the package.json upgrades. It's too many changes for this PR and maybe its one of them that breaks the whole thing.

sallar commented:
@perry-mitchell @ph1p done!

ph1p commented:
Thanks @sallar. This was a stupid idea of mine, to push all code into the master branch 🤗. Next time I’ll create an extra feature branch and another pull request.

perry-mitchell commented:
> This was a stupid idea of mine

No such thing! This was a huge chunk of work. We obviously need to update packages anyway.. 😁

ph1p commented:
Ok I've one last commit with some bugfixes and translations (no dependency updates). In about 20 minutes I'll push it.

sallar commented:
Great work @ph1p. I'm going to merge this :)
